File: example-jv-01reactiveui/src/main/java/foo/bar/example/forereactiveui/ui/wallet/WalletsActivity.java
package foo.bar.example.forereactiveui.ui.wallet;

import android.os.Bundle;
import android.widget.Button;
import android.widget.TextView;

import androidx.fragment.app.FragmentActivity;
import butterknife.BindView;
import butterknife.ButterKnife;
import co.early.fore.core.observer.Observer;
import foo.bar.example.forereactiveui.OG;
import foo.bar.example.forereactiveui.R;
import foo.bar.example.forereactiveui.feature.wallet.Wallet;


public class WalletsActivity extends FragmentActivity {

    //models that we need to sync with
    private Wallet wallet = OG.get(Wallet.class);


    //UI elements that we care about
    @BindView(R.id.wallet_increase_btn)
    public Button increaseMobileWalletBtn;

    @BindView(R.id.wallet_decrease_btn)
    public Button decreaseMobileWalletBtn;

    @BindView(R.id.wallet_mobileamount_txt)
    public TextView mobileWalletAmount;

    @BindView(R.id.wallet_savingsamount_txt)
    public TextView savingsWalletAmount;


    //single observer reference
    Observer observer = this::syncView;



    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);

        setContentView(R.layout.activity_wallet);

        ButterKnife.bind(this);

        setupButtonClickListeners();
    }

    private void setupButtonClickListeners() {
        increaseMobileWalletBtn.setOnClickListener(v -> {
            wallet.increaseMobileWallet();//notice how the data binding takes care of updating the view for you
        });
        decreaseMobileWalletBtn.setOnClickListener(v -> {
            wallet.decreaseMobileWallet();//notice how the data binding takes care of updating the view for you
        });
    }


    //reactive UI stuff below

    public void syncView(){
        increaseMobileWalletBtn.setEnabled(wallet.canIncrease());
        decreaseMobileWalletBtn.setEnabled(wallet.canDecrease());
        mobileWalletAmount.setText("" + wallet.getMobileWalletAmount());
        savingsWalletAmount.setText("" + wallet.getSavingsWalletAmount());
    }

    @Override
    protected void onStart() {
        super.onStart();
        wallet.addObserver(observer);
        syncView(); //  <- don't forget this
    }


    @Override
    protected void onStop() {
        super.onStop();
        wallet.removeObserver(observer);
    }

}

File: example-kt-01reactiveui/src/main/java/foo/bar/example/forereactiveuikt/ui/wallet/WalletsActivity.kt
package foo.bar.example.forereactiveuikt.ui.wallet

import android.os.Bundle
import androidx.fragment.app.FragmentActivity
import co.early.fore.core.observer.Observer
import co.early.fore.core.ui.SyncableView
import foo.bar.example.forereactiveuikt.OG
import foo.bar.example.forereactiveuikt.R
import foo.bar.example.forereactiveuikt.feature.wallet.Wallet
import kotlinx.android.synthetic.main.activity_wallet.*

/**
 * Copyright © 2015-2020 early.co. All rights reserved.
 */
class WalletsActivity : FragmentActivity(R.layout.activity_wallet), SyncableView {

    //models that we need to sync with
    private val wallet: Wallet = OG[Wallet::class.java]

    //single observer reference
    var observer = Observer { syncView() }


    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)

        setupButtonClickListeners()
    }

    private fun setupButtonClickListeners() {
        wallet_increase_btn.setOnClickListener {
            wallet.increaseMobileWallet() //notice how the reactive ui takes care of updating the view for you
        }
        wallet_decrease_btn.setOnClickListener {
            wallet.decreaseMobileWallet() //notice how the reactive ui takes care of updating the view for you
        }
    }

    //reactive UI stuff below
    override fun syncView() {
        wallet_increase_btn.isEnabled = wallet.canIncrease()
        wallet_decrease_btn.isEnabled = wallet.canDecrease()
        wallet_mobileamount_txt.text = wallet.mobileWalletAmount.toString()
        wallet_savingsamount_txt.text = wallet.savingsWalletAmount.toString()
    }

    override fun onStart() {
        super.onStart()
        wallet.addObserver(observer)
        syncView() //  <- don't forget this
    }

    override fun onStop() {
        super.onStop()
        wallet.removeObserver(observer)
    }
}
